### PR TITLE
Implement Closures

### LINKED ISSUE #566: Implement Lambdas
# Problem

After #468 is merged, it follows Noir should also support unnamed functions - lambdas. Lambdas will make it easier to use first class functions by removing the requirement each function be named and defined globally. This is much more ergonomic for small functions.

# Solution

Add lambdas to noir, copying rust's syntax sans `move` keyword.

```rs
fn main() {
    let array1 = [1, 2, 3];
    let array2 = map(array1, 0, |x| x + 1);
    constrain array2 == [2, 3, 4];
}

// map an array to an array with f applied to each element.
// need `initial` currently to initialize the second array.
fn map<T, U>(array: [T], initial: U, f: fn(T) -> U) -> [U] {
    let len = std::array::len(array);
    let mut ret = [initial; len];
    for i in 0 .. len {
        ret[i] = f(array[i]);
    }
    ret
}
```

### PR DESCRIPTION
**Pull request by jfecher**
# Related issue(s)

Resolves #566

# Description

## Summary of changes

Implements lambdas and closures, allowing anonymous functions to be declared and used. These closures can also capture and use any contained variables of the parent function's scope they were declared in.

Closures are limited currently to only capture immutable variables due to a bug when capturing mutable variables. We can remove this restriction in the future but for now capturing immutable variables is very flexible since you can always copy a mutable variable into an immutable one and capture that.

~~This PR requires the higher order functions PR (#468) to work and should only be merged after it is.~~

## Test additions / changes

Adds lambdas and closures to the higher_order_functions test

# Checklist

- [x] I have tested the changes locally.
- [x] I have formatted the changes with [Prettier](https://prettier.io/) and/or `cargo fmt` with default settings.
- [x] I have [linked](https://docs.github.com/en/issues/tracking-your-work-with-issues/linking-a-pull-request-to-an-issue) this PR to the issue(s) that it resolves.
- [x] I have reviewed the changes on GitHub, line by line.
- [x] I have ensured all changes are covered in the description.

### CODE REVIEW

**Review by vezenovm — COMMENTED** [review]
Minor nitpicks but it looks good

File: crates/noirc_frontend/src/monomorphisation/mod.rs [hunk 1]
```diff
@@ -473,6 +498,7 @@ impl Monomorphiser {
 
     /// Convert a non-tuple/struct type to a monomorphised type
     fn convert_type(typ: &HirType) -> ast::Type {
+        println!("Converting type {}", typ);
```
[the remark this answers is not in the document]
Comment (jfecher) [reply crates/noirc_frontend/src/monomorphisation/mod.rs:501]: Ack, good catch.

**Review by guipublic — APPROVED** [review]
Fine for me.

**Comment by vezenovm** [comment]
Can merge once CI is fixed.

We seem to keep running into this causing our CI to fail:
```
error: duplicate definitions of hello function found
  ┌─ /home/runner/work/noir/noir/crates/nargo/tests/compile_tests_data/fail/dup_func.nr:2:4
  │
2 │ fn hello(x : Field) -> Field {
  │    ----- first definition found here
  ·
6 │ fn hello(x : Field) -> Field {
  │    ----- second definition found here
```
If you re-run the job for now it seems to always resolve, but we should fix this so it works on the first CI run.

**Comment by jfecher** [comment]
> We seem to keep running into this causing our CI to fail

That is not the error, that is an error from a test case that is expected to fail. The only reason that output is being shown is because when there is an error further down the line it triggers all output to be shown rather than hidden. The actual error causing the tests to fail is the `SIGILL: Illegal Instruction` you can see further down

**Comment by vezenovm** [comment]
> The actual error causing the tests to fail is the `SIGILL: Illegal Instruction` you can see further down.

Ah I missed that was the cause of that output, thanks for clarifying. So this is the same error the CI has been intermittently running into.

**Pull request merged**